src/merkle/tiered_smt/nodes.rs

```diff
@@ -0,0 +1,365 @@
+        let mse = get_key_prefix(key);
+        for depth in (TIER_DEPTHS[0]..MAX_DEPTH).step_by(TIER_SIZE as usize) {
+            let index = NodeIndex::new_unchecked(depth, mse >> (MAX_DEPTH - depth));
+            if self.upper_leaves.contains(&index) {
+                return (index, true);
+            } else if !self.nodes.contains_key(&index) {
+                return (index, false);
+            }
+        }
+
+        // if we got here, that means all of the nodes checked so far are internal nodes, and
+        // the new node would need to be inserted in the bottom tier.
+        let index = NodeIndex::new_unchecked(MAX_DEPTH, mse);
+        (index, false)
+    }
+
+    // ITERATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns an iterator over all inner nodes of the Tiered Sparse Merkle tree (i.e., nodes not
+    /// at depths 16 32, 48, or 64).
+    ///
+    /// The iterator order is unspecified.
+    pub fn inner_nodes(&self) -> impl Iterator<Item = InnerNodeInfo> + '_ {
+        self.nodes.iter().filter_map(|(index, node)| {
+            if self.is_internal_node(index) {
+                Some(InnerNodeInfo {
+                    value: *node,
+                    left: self.get_node_unchecked(&index.left_child()),
+                    right: self.get_node_unchecked(&index.right_child()),
+                })
+            } else {
+                None
+            }
+        })
+    }
+
+    /// Returns an iterator over the upper leaves (i.e., leaves with depths 16, 32, 48) of the
+    /// Tiered Sparse Merkle tree.
+    pub fn upper_leaves(&self) -> impl Iterator<Item = (&NodeIndex, &RpoDigest)> {
+        self.upper_leaves.iter().map(|index| (index, &self.nodes[index]))
+    }
+
+    /// Returns an iterator over the bottom leaves (i.e., leaves with depth 64) of the Tiered
+    /// Sparse Merkle tree.
+    pub fn bottom_leaves(&self) -> impl Iterator<Item = (&u64, &RpoDigest)> {
+        self.bottom_leaves.iter().map(|value| {
+            let index = NodeIndex::new_unchecked(MAX_DEPTH, *value);
+            (value, &self.nodes[&index])
+        })
+    }
+
+    // STATE MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Replaces the leaf node at the specified index with a tree consisting of two leaves located
+    /// at the specified indexes. Recomputes and returns the new root.
+    pub fn replace_leaf_with_subtree(
+        &mut self,
+        leaf_index: NodeIndex,
+        subtree_leaves: [(NodeIndex, RpoDigest); 2],
+    ) -> RpoDigest {
+        debug_assert!(is_tiered_node(&leaf_index));
+        debug_assert!(is_tiered_node(&subtree_leaves[0].0));
+        debug_assert!(is_tiered_node(&subtree_leaves[1].0));
+        debug_assert!(!is_empty_root(&subtree_leaves[0].1));
+        debug_assert!(!is_empty_root(&subtree_leaves[1].1));
```

Comment:
> doesn't this prevent a user from assigning a tree as a leaf value?

Yeah - this could happen, but a check to catch this would be fairly complicated (e.g., we'd need to pass the actual key-value pairs, hash them, and compute node values from them) - so, I'm not sure it is worth it.

Reply:
I was proposing removing the existing check, not adding another one. The assumption is that hash collisions won't happen because of the use of `domain` in the hash function.